subscriber/common/job/controller.go

```diff
@@ -517,6 +507,34 @@ func (c *Controller) startDriver(
 	return true
 }
 
+func (c *Controller) startEtcdHBService(params Params) {
+	var err error
+	params.ServiceConfig.Logger.Info("heartbeat config",
+		zap.Any("heartbeat", *params.ServiceConfig.HeartbeatConfig))
+	params.ServiceConfig.EtcdConfig.Lock()
+	c.etcdServices, err = connectEtcdServices(params)
+	if err != nil {
+		params.ServiceConfig.Logger.Panic("Failed to createEtcdServices", zap.Error(err))
+	}
+	if registerHeartBeatService(params, c.etcdServices) != nil {
+		params.ServiceConfig.Logger.Panic("Failed to registerHeartBeatService", zap.Error(err))
+	}
+	params.ServiceConfig.EtcdConfig.Unlock()
+}
+
+// RestartEtcdHBService registers heartbeat again if etcd cluster changes are detected
+func (c *Controller) RestartEtcdHBService(params Params) {
+	for {
+		select {
+		case <-config.EtcdCfgEvent:
+			// TODO: unadevertises old heartbeat and closes etcd client should be added once M3 provides
+			c.serviceConfig.Logger.Info("RestartEtcdHBService")
+			c.startEtcdHBService(params)
+		default:
```

Comment:
having default here turns this into a busy loop as opposed to go routine put into sleep until etcd event actually arrive

Reply:
done